boto/s3/bucket.py

```diff
@@ -879,15 +884,33 @@ def delete_keys2(headers):
             data += u"</Delete>"
             if count <= 0:
                 return False  # no more
-            data = data.encode('utf-8')
-            fp = BytesIO(data)
-            md5 = boto.utils.compute_md5(fp)
+            md5 = boto.utils.compute_md5(StringIO(data))
             headers['Content-MD5'] = md5[1]
             headers['Content-Type'] = 'text/xml'
+            if not isinstance(data, bytes):
+                data = data.encode('utf-8')
             if mfa_token:
                 headers[provider.mfa_header] = ' '.join(mfa_token)
             if bypass_governance_retention is not None:
-                headers[self.connection.provider.bypass_governance_retention_header] = bypass_governance_retention
+                headers[provider.bypass_governance_retention_header] = bypass_governance_retention
+            if checksum is not None:
+                # "x-amz-sdk-checksum-algorithm" header: CRC32|CRC32C|SHA1|SHA256
+                # If key/value already provided in headers dictionary, we will use it as is.
+                if provider.sdk_checksum_algorithm_header not in headers:
+                    headers[provider.sdk_checksum_algorithm_header] = checksum
+                # calculate based on the algorithm
+                checksum_lower = checksum.lower()
+                if checksum_lower in ['crc32', 'crc32c', 'sha1', 'sha256']:
+                    calculated_checksum = cal_checksum(None, -1, checksum_lower, data)
+                    # If key/value already provided in headers dictionary, we will use it as is.
+                    if checksum_lower == 'crc32' and provider.checksum_crc32_header not in headers:
+                        headers[provider.checksum_crc32_header] = calculated_checksum
+                    elif checksum_lower == 'crc32c' and provider.checksum_crc32c_header not in headers:
+                        headers[provider.checksum_crc32c_header] = calculated_checksum
+                    elif checksum_lower == 'sha1' and provider.checksum_sha1_header not in headers:
+                        headers[provider.checksum_sha1_header] = calculated_checksum
+                    elif checksum_lower == 'sha256' and provider.checksum_sha256_header not in headers:
+                        headers[provider.checksum_sha256_header] = calculated_checksum
```

Comment:
This is just copy and paste for many of the function. It cries out to be an internal helper function don't you think?

Reply:
Added a new func as  **_configure_checksum_headers_** method in checksums.py and overall simplified.